stl/src/print.cpp

```diff
@@ -0,0 +1,280 @@
+    public:
+        _Transcode_result() noexcept : _Transcoded_str(), _Successful(true) {}
+
+        _Transcode_result(_Minimal_wstring_view _Result_str) noexcept
+            : _Transcoded_str(_Result_str), _Successful(true) {}
+
+        _Transcode_result(__std_win_error _Result_error) noexcept : _Win_error(_Result_error), _Successful(false) {}
+
+        [[nodiscard]] bool _Has_value() const noexcept {
+            return _Successful;
+        }
+
+        [[nodiscard]] _Minimal_wstring_view _Value() const noexcept {
+            return _Transcoded_str;
+        }
+
+        [[nodiscard]] __std_win_error _Error() const noexcept {
+            return _Win_error;
+        }
+
+    private:
+        union {
+            _Minimal_wstring_view _Transcoded_str;
+            __std_win_error _Win_error;
+        };
+
+        bool _Successful;
+    };
+
+    [[nodiscard]] _Transcode_result _Transcode_utf8_string(
+        _Allocated_string& _Dst_str, const _Minimal_string_view _Src_str) noexcept {
+        // MultiByteToWideChar() fails if strLength == 0.
+        if (_Src_str._Empty()) [[unlikely]] {
+            return {};
+        }
+
+        // For vprint_unicode(), N4928 [ostream.formatted.print]/4 suggests replacing invalid code units with U+FFFD.
+        // This is done automatically by MultiByteToWideChar(), so long as we do not use the MB_ERR_INVALID_CHARS flag.
+        // We transcode up to 8,192 bytes per segment, which easily fits in an int.
+        const int32_t _Num_chars_required =
+            MultiByteToWideChar(CP_UTF8, 0, _Src_str._Data(), static_cast<int>(_Src_str._Size()), nullptr, 0);
+
+        if (_Num_chars_required == 0) [[unlikely]] {
+            return static_cast<__std_win_error>(GetLastError());
+        }
+
+        if (static_cast<size_t>(_Num_chars_required) > _Dst_str._Capacity()) {
+            _Dst_str._Reset();
+
+            __crt_unique_heap_ptr<wchar_t> _Wide_str{_malloc_crt_t(wchar_t, _Num_chars_required)};
+            if (!_Wide_str) [[unlikely]] {
+                return __std_win_error::_Not_enough_memory;
+            }
+
+            _Dst_str = _Allocated_string{_STD move(_Wide_str), static_cast<size_t>(_Num_chars_required)};
+        }
+
+        const int32_t _Conversion_result = MultiByteToWideChar(CP_UTF8, 0, _Src_str._Data(),
+            static_cast<int>(_Src_str._Size()), _Dst_str._Data(), static_cast<int>(_Dst_str._Capacity()));
+
+        if (_Conversion_result == 0) [[unlikely]] {
+            // This shouldn't happen...
+            _CSTD abort();
+        }
+
+        return _Minimal_wstring_view{_Dst_str._Data(), static_cast<size_t>(_Conversion_result)};
+    }
+
+    [[nodiscard]] __std_win_error _Write_console(
+        const HANDLE _Console_handle, const _Minimal_wstring_view _Wide_str) noexcept {
+        const BOOL _Write_result =
+            WriteConsoleW(_Console_handle, _Wide_str._Data(), static_cast<DWORD>(_Wide_str._Size()), nullptr, nullptr);

```

Comment:
[`GetConsoleMode()`](https://learn.microsoft.com/en-us/windows/console/getconsolemode) and [`WriteConsoleW()`](https://learn.microsoft.com/en-us/windows/console/writeconsole) are documented as "desktop apps only". Do we need to fail for `_CRT_APP`?

Reply:
Under what conditions is the `_CRT_APP` macro defined? Is that for UWP apps? I can't seem to find anything about it online, and it doesn't seem to be set in the `CMakeLists.txt` file, so I assume that it's an internal macro.

Reply:
This was discussed offline. While their individual documentation pages state as such, it's because they are outdated. Both APIs are listed as allowed on [this page](https://learn.microsoft.com/en-us/uwp/win32-and-com/win32-apis#apis-from-api-ms-win-core-console-l1-1-0dll) and the headers also provide them under UWP. So there shouldn't be anything to fix here.

Reply:
Yeah, `_CRT_APP` is an internal macro defined by the legacy build system:
```
<ClDefines Condition="'$(TargetAppFamily)' == 'true'">$(ClDefines);_CRT_APP</ClDefines>
```
Thanks @sylveon for investigating this - I agree that no action should be required. If we discover app problems, I believe we can patch them without too much trouble.